Add main menu for Core Sample Scenes and QA Tests, Minor RebindingActionUI script changes - ISX-1776

Pauliusd01 opened the pull request:
### Description

Minor changes to our main repo scenes for better testing quality of life as well as user requested changes to the Rebinding UI Sample. Example of the scene loader: 
https://github.com/Unity-Technologies/InputSystem/assets/54306142/029a3580-9b05-431a-ab8a-b018a041f891
Works, or at least should work both with Keyboard&Mouse and Controller cursor. (Esc and Start button bring up the menu to return to the loader)

### Changes made

- Adds user suggested changes to the RebindingUI sample (https://forum.unity.com/threads/rebinding-ui-sample-bugs-fixes.1307430/)
- Adds a main menu scene with a scene loader for easier access to the core platforms so you no longer need to rebuild the player to change/reset scenes
- Moved DocExamples scenes by Ben to QA folder as they are incomplete and not publicly accessible samples
- Some scenes had their event system objects updated to use the IS version. Seemed fine as the ones I changed did not mention to be Input Manager related.

### Notes

I've also noticed that there might be a bug with ProjectWideAction references when loading/unloading scenes: 
https://github.com/Unity-Technologies/InputSystem/assets/54306142/c39eda98-5b11-4a4b-99cc-dd002ae00085
Would like to report this separately for the developers to investigate after these changes merge.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

Pauliusd01 committed:
RebindActionUI changes, SceneLoader

Pauliusd01 committed:
Delete Assets/QA/Input_Test/Editor/default.preset

lyndon-unity commented on Assets/AddBuildScenes.cs:
```diff
@@ -0,0 +1,31 @@
+        List<SceneAsset> m_SceneAssets = new List<SceneAsset>();
+        // Find valid Scene paths and make a list of EditorBuildSettingsScene
+        List<EditorBuildSettingsScene> editorBuildSettingsScenes = new List<EditorBuildSettingsScene>();
+        foreach (var sceneAsset in m_SceneAssets)
```
Looks like this doesn't do anything - m_SceneAssets will be an empty list as its only just created.

Pauliusd01 replied:
Ah, this one wasn't supposed to be here at all. First attempt at it

Pauliusd01 committed:
Update AddBuildScenes.cs

lyndon-unity commented on Assets/QA/Tests/Core Platform Menu/AddScenesToBuild.cs:
```diff
@@ -0,0 +1,108 @@
+    {
+        if (state == PlayModeStateChange.ExitingPlayMode)
+        {
+            // Save the build settings before entering play mode
```
Isn't this the wrong way round. Sounds like its exiting play mode and we wouldn't want to be saving during that phase.

Pauliusd01 replied:
Leaving play mode in Editor would always clear my added build scenes so the next time I enter play mode I couldn't use the scene loader. Let me see if the issue is still there and what I can do

lyndon-unity commented on Assets/Samples/RebindingUI/RebindActionUI.cs:
```diff
@@ -268,14 +268,16 @@ void CleanUp()
                     operation =>
                     {
                         m_RebindStopEvent?.Invoke(this, operation);
-                        m_RebindOverlay?.SetActive(false);
```
Why this change - its the same thing isn't it.

Pauliusd01 replied:
Changes suggested by a user: https://forum.unity.com/threads/rebinding-ui-sample-bugs-fixes.1307430/

"The Elvis operator doesn't work with UnityEngine.Object types in some circumstances. It seems to be a long standing issue, see [this thread](https://forum.unity.com/threads/elvis-operator-null-coalescing-operator.433767/) for details."

lyndon-unity reviewed (changes requested):
Added some comments as I'm not sure some of the changes are valid

Pauliusd01 committed:
Update Scene Loader.cs

Pauliusd01 committed:
Delete Assets/AddBuildScenes.cs

Pauliusd01 committed:
Delete Assets/AddBuildScenes.cs.meta

Pauliusd01 committed:
Move Scene adder to Editor folder, add buttons for more scenes to loader

Pauliusd01 committed:
Delete Assets/QA/Input_Test/Editor/default.preset.meta

Pauliusd01 committed:
Update event system object for DeviceList scene

Pauliusd01 committed:
Move DocsExamples to QA folder

Moved doc examples to QA folder as they are publicly available samples. Intention is to add changes to them and use them for QA purposes instead

Pauliusd01 committed:
Update Core Platforms Menu.unity

Pauliusd01 committed:
Update RebindActionUI.cs

Pauliusd01 committed:
Update AddScenesToBuild.cs

Pauliusd01 committed:
Update Scene Loader.cs

fix formatting

Pauliusd01 committed:
formatting

Pauliusd01 committed:
formatting fix

Pauliusd01 committed:
added more scenes to the loader, changed event system components for old scenes

lyndon-unity commented on Assets/QA/Input_Test/IMETest.unity:
There are a number of changed .unity files which seem just related to upgrading to a newer unity version. I'm not sure this need to be in the PR and may prevent running tests on older Unity versions. Can these be removed. 

Likewise some unneccessary .meta file changes

Pauliusd01 replied:
They've actually been modified intentionally by this:
>Some scenes had their event system objects updated to use the IS version. Seemed fine as the ones I changed did not mention to be Input Manager related.

If I don't make these changes those scenes will throw an error on load.

>Likewise some unneccessary .meta file changes

Can you point out which ones? I don't see any loose .meta files. Unless you mean the changes to the scenes again and I do agree that they do come with more than just that EventSystem component changes. I think I could ditch those although doesn't seem obvious which parts to keep.

ekcoh replied:
I would advise to make any of those upgrades to .unity files and meta files using Unity 2019 since that is the min version required by Input System 1.x and hence should verify that there are no issues with 2019 due to any files being in a newer format (avoid downgrade scenarios). I believe comment by @lyndon-unity regarding .meta file changes is any of the diffs on the PR relating to already existing .meta files being modified but I couldn't spot any such changes in the diff so I let @lyndon-unity provide additional detail regarding that.

Pauliusd01 replied:
That is a good suggestion, I will redo those files using 2019.4

lyndon-unity reviewed (changes requested):
General concept of adding ui menu and moving some manual tests over to the QA area seems reasonable - comment about upgraded .unity files that seems risky if changes not needed.

ekcoh reviewed (commented):
Did you confirm with @duckets that the move of DocExamples is purely structural and do not have any impact on doc generation or publishing?

Pauliusd01 commented:
> Did you confirm with @duckets that the move of DocExamples is purely structural and do not have any impact on doc generation or publishing?

Yes, he even suggested deleting them but I thought It's better to keep and modify as testing scenes for internal use

Pauliusd01 committed:
Delete Assets/QA/Input_Test/IMETest.unity

Pauliusd01 committed:
Revert "Delete Assets/QA/Input_Test/IMETest.unity"

This reverts commit c07d99a823403b3e5b60481de53b9fdc8d9dd0ab.

Pauliusd01 committed:
Revert "changed event system components for old scenes"

This reverts commit f78fcde6b8b1ed34058994ab31364aa05124c4b5.

Pauliusd01 committed:
Revert DeviceList.Unity changes

Pauliusd01 committed:
Revert EditModeTest.unity changes

Pauliusd01 committed:
Changed event system components for old scenes to IS compatible one (using 2019.4)

Pauliusd01 committed:
Update IMETest.unity

Pauliusd01 committed:
Revert "Update IMETest.unity"

This reverts commit debf260e759430b08ce3666b79ec42ae60388e29.

Pauliusd01 committed:
Revert "Changed event system components for old scenes to IS compatible one (using 2019.4)"

This reverts commit 69fb375e787d8254c0e82ea566691301dc689158.

Pauliusd01 committed:
Update IMETest.unity

Pauliusd01 committed:
Update Core Platforms Menu.unity

Pauliusd01 committed:
Update DeviceList event system to IS compatible (2019.4)

Pauliusd01 committed:
Update ISX_KB_Mouse_Basic event system to IS compatible (2019.4)

Pauliusd01 committed:
Update ISX_Touch event system to IS compatible (2019.4)

Pauliusd01 committed:
Merge branch 'develop' into update-sample-scenes-ISX-1776

Pauliusd01 commented:
@lyndon-unity @ekcoh I've reverted my event system changes to the old scenes and redone them again in 2019.4. I hope that feels safer than the previous attempt

lyndon-unity commented on Assets/Editor/AddScenesToBuild.cs:
```diff
@@ -0,0 +1,90 @@
+        }
+
+        // Ensure "Core Platforms Menu" is at the beginning of the list
+        filteredScenePaths.Insert(0, "Assets/QA/Tests/Core Platform Menu/Core Platforms Menu.unity");
```
I was wondering if this gets added twice as scenePaths would pickup the scene too ?

ekcoh replied:
agreed, would recommend to either skip/filter it out in the foreach above and add it like this or detect in within the loop and use Insert(0 instead of the Add for that particular file.

lyndon-unity reviewed (approved):
This looks much better being based on 2019 Unity for the .unity / .meta files

ekcoh commented on Assets/Editor/AddScenesToBuild.cs:
```diff
@@ -0,0 +1,90 @@
+    [MenuItem("QA Tools/Open Core Scene Menu")]
+    static void OpenScene()
+    {
+        EditorSceneManager.OpenScene("Assets/QA/Tests/Core Platform Menu/Core Platforms Menu.unity");
```
This exact path is also present on line 37. I Would suggest you put it into a constant e.g.

`private const string corePlatformsMenu = "Assets/QA/Tests/Core Platform Menu/Core Platforms Menu.unity"`

and then reference it from all places used.

ekcoh reviewed (approved):
Approved, but would suggest switching to a constant for a string used more than once as well as address the comment by Lyndon about added twice.

ekcoh commented:
Needs to be updated with latest changes from develop and CI rerun before merge.

lyndon-unity committed:
Merge branch 'develop' into update-sample-scenes-ISX-1776

Pauliusd01 committed:
Update Scene adding to skip menu scene and insert it later

Pauliusd01 committed:
Merge branch 'develop' into update-sample-scenes-ISX-1776
